File: backend/controllers/farm_controller.go
package controllers
import (
	"net/http"
	"cronos-backend/middleware"
	"cronos-backend/models"
	"cronos-backend/services"
	"cronos-backend/utils"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)
type FarmController struct {
	FarmService *services.FarmService
}
func NewFarmController(farmService *services.FarmService) *FarmController {
	return &FarmController{FarmService: farmService}
}
func (ctrl *FarmController) CreateFarm(c *gin.Context) {
	userID := middleware.GetUserID(c)
	var farm models.Farm
	if err := c.ShouldBindJSON(&farm); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	if err := ctrl.FarmService.CreateFarm(userID, &farm); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusCreated, "Farm created successfully", farm)
}
func (ctrl *FarmController) GetMyFarms(c *gin.Context) {
	userID := middleware.GetUserID(c)
	farms, err := ctrl.FarmService.GetMyFarms(userID)
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Farms retrieved", farms)
}
func (ctrl *FarmController) GetFarm(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid farm ID", nil)
		return
	}
	farm, err := ctrl.FarmService.GetFarm(id)
	if err != nil {
		utils.NotFound(c, "Farm not found")
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Farm retrieved", farm)
}
func (ctrl *FarmController) UpdateFarm(c *gin.Context) {
	userID := middleware.GetUserID(c)
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid farm ID", nil)
		return
	}
	var req models.Farm
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	farm, err := ctrl.FarmService.UpdateFarm(userID, id, &req)
	if err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Farm updated", farm)
}
func (ctrl *FarmController) DeleteFarm(c *gin.Context) {
	userID := middleware.GetUserID(c)
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid farm ID", nil)
		return
	}
	if err := ctrl.FarmService.DeleteFarm(userID, id); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Farm deleted", nil)
}
func (ctrl *FarmController) CreatePond(c *gin.Context) {
	userID := middleware.GetUserID(c)
	farmID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid farm ID", nil)
		return
	}
	var pond models.Pond
	if err := c.ShouldBindJSON(&pond); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	if err := ctrl.FarmService.CreatePond(userID, farmID, &pond); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusCreated, "Pond created", pond)
}
func (ctrl *FarmController) GetPonds(c *gin.Context) {
	farmID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid farm ID", nil)
		return
	}
	ponds, err := ctrl.FarmService.GetPonds(farmID)
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Ponds retrieved", ponds)
}
func (ctrl *FarmController) UpdatePond(c *gin.Context) {
	userID := middleware.GetUserID(c)
	pondID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid pond ID", nil)
		return
	}
	var req models.Pond
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	pond, err := ctrl.FarmService.UpdatePond(userID, pondID, &req)
	if err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Pond updated", pond)
}
func (ctrl *FarmController) DeletePond(c *gin.Context) {
	userID := middleware.GetUserID(c)
	pondID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid pond ID", nil)
		return
	}
	if err := ctrl.FarmService.DeletePond(userID, pondID); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Pond deleted", nil)
}

File: backend-spring/src/main/java/com/cronos/controller/FarmController.java
package com.cronos.controller;

import com.cronos.dto.response.ApiResponse;
import com.cronos.entity.Farm;
import com.cronos.entity.Pond;
import com.cronos.service.FarmService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.UUID;

@RestController
@RequestMapping("/api/farms")
@RequiredArgsConstructor
public class FarmController {
    private final FarmService farmService;

    @PostMapping
    public ResponseEntity<?> create(Authentication auth, @RequestBody Farm farm) {
        UUID userId = (UUID) auth.getPrincipal();
        return ResponseEntity.ok(ApiResponse.success("Farm created", farmService.createFarm(userId, farm)));
    }

    @GetMapping
    public ResponseEntity<?> getMyFarms(Authentication auth) {
        UUID userId = (UUID) auth.getPrincipal();
        return ResponseEntity.ok(ApiResponse.success("Farms retrieved", farmService.getMyFarms(userId)));
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getFarm(@PathVariable UUID id) {
        return ResponseEntity.ok(ApiResponse.success("Farm retrieved", farmService.getFarm(id)));
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> update(Authentication auth, @PathVariable UUID id, @RequestBody Farm req) {
        UUID userId = (UUID) auth.getPrincipal();
        try {
            return ResponseEntity.ok(ApiResponse.success("Farm updated", farmService.updateFarm(userId, id, req)));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> delete(Authentication auth, @PathVariable UUID id) {
        UUID userId = (UUID) auth.getPrincipal();
        try {
            farmService.deleteFarm(userId, id); return ResponseEntity.ok(ApiResponse.success("Farm deleted", null));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @PostMapping("/{farmId}/ponds")
    public ResponseEntity<?> createPond(Authentication auth, @PathVariable UUID farmId, @RequestBody Pond pond) {
        UUID userId = (UUID) auth.getPrincipal();
        try {
            return ResponseEntity.ok(ApiResponse.success("Pond created", farmService.createPond(userId, farmId, pond)));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @GetMapping("/{farmId}/ponds")
    public ResponseEntity<?> getPonds(@PathVariable UUID farmId) {
        return ResponseEntity.ok(ApiResponse.success("Ponds retrieved", farmService.getPonds(farmId)));
    }

    @PutMapping("/ponds/{pondId}")
    public ResponseEntity<?> updatePond(Authentication auth, @PathVariable UUID pondId, @RequestBody Pond req) {
        UUID userId = (UUID) auth.getPrincipal();
        try {
            return ResponseEntity.ok(ApiResponse.success("Pond updated", farmService.updatePond(userId, pondId, req)));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @DeleteMapping("/ponds/{pondId}")
    public ResponseEntity<?> deletePond(Authentication auth, @PathVariable UUID pondId) {
        UUID userId = (UUID) auth.getPrincipal();
        try {
            farmService.deletePond(userId, pondId); return ResponseEntity.ok(ApiResponse.success("Pond deleted", null));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }
}
